README.rst
Python Atomx Api
================

Interface for the atomx rest api.

For more information read the full
`documentation online <http://atomx-api-python.readthedocs.org/en/latest/index.html>`_,
report bugs in `github <https://github.com/atomx/atomx-api-python>`_
or see the `atomx wiki <https://wiki.atomx.com/api>`_


Example Usage
-------------

.. code-block:: python

    from atomx import Atomx

    # create atomx session
    atomx = Atomx('user@example.com', 'password')

    # get 10 creatives
    creatives = atomx.get('Creatives', limit=10)
    # the result is a list of `atomx.models.Creative` models
    # that you can easily inspect, manipulate and update
    for creative in creatives:
        print('Creative ID: {c.id}, state: {c.state}, '
              'name: {c.name}, title: {c.title}'.format(c=creative))

    # update title for the first creative in list
    creative = creatives[0]
    creative.title = 'shiny new title'
    # the session is inherited from `atomx` that made the get request
    creative.save()


    # create a new profile
    from atomx.models import Profile
    profile = Profile(advertiser_id=23, name='test profile')
    # Note that you have to pass it a valid `Atomx` session for create
    # or use `atomx.create(profile)`
    profile.create(atomx)

    # now you could alter and update it like the creative above
    profile.name = 'changed name'
    profile.save()


    # you can also get attributes
    profiles = atomx.get('advertiser', 88, 'profiles')
    # equivalent is to pass the complete resource path as string instead of arguments
    profiles = atomx.get('advertiser/88/profiles')  # same as above
    # profiles is now a list of `atomx.models.Profile` that you can
    # read, update, etc again.
    profiles[0].click_frequency_cap_per = 86400
    profiles[0].save()


    # working with search
    s = atomx.search('mini*')
    # s is now a dict with lists of search results for the different models
    # with the model id and name

    publisher = s['publisher'][0]  # get the first publisher..
    publisher.reload()  # .. and load all the data
    print(publisher)  # now all publisher data is there
    publisher.history()  # gets all changes made to this publisher


    # reporting example
    # get a report for a specific publisher
    report = atomx.report(scope='publisher', groups=['hour'], metrics=['impressions', 'clicks'], where=[['publisher_id', '==', 42]], from_='2015-02-08 00:00:00', to='2015-02-09 00:00:00', timezone='America/Los_Angeles')
    # check if report is ready
    print(report.is_ready)
    # if pandas is installed you can get the pandas dataframe with `report.pandas`
    # you can also get the report csv in `report.content` without pandas
    df = report.pandas  # A datetime index is automatically set when group by a hour/day/month.
    # calculate mean, median, std per hour
    means = df.resample('H', how=['mean', 'median', 'std'])
    # and plot impression and clicks per day
    means['impressions'].plot()
    means['clicks'].plot()


Installation
------------

To install the python atomx api, simply:

.. code-block:: bash

    $ pip install atomx

or if you want to use ipython notebook and reporting functionality:

.. code-block:: bash

    $ pip install atomx[report]


CHANGES.txt
1.7
---

- Add ``editable`` parameter to :meth:`atomx.Atomx.report`.
- Add ``dsp`` reporting scope.
- Add ``daterange`` parameter to :meth:`atomx.Atomx.report`.
- When accessing a model attribute, attributes that are atomx models will now
  automatically returned as :class:`atomx.models.AtomxModel` instead of
  leaving them as `dict`.
- :meth:`atomx.models.AtomxModel.reload` takes kwargs parameter
- :meth:`atomx.Atomx.delete` can take a `atomx.models` instance as argument
- Add new models:

  * :class:`atomx.models.Dsp`
  * :class:`atomx.models.Ssp`
  * :class:`atomx.models.SspSuspicious`
  * :class:`atomx.models.SspResultType`


1.6
---

- :meth:`atomx.Atomx.login` takes ``totp`` parameter for users that have 2-factor auth enabled.
- :meth:`atomx.Atomx.search` takes ``index`` parameter to only search specific models.
- Add support to remove models from the API. See :meth:`atomx.models.AtomxModel.delete`.
- Add :meth:`atomx.Atomx.remove`.
- Add ``save`` parameter to :meth:`atomx.Atomx.report`.
- Add new models:

  * :class:`atomx.models.App`
  * :class:`atomx.models.Appstore`
  * :class:`atomx.models.City`
  * :class:`atomx.models.Dma`
  * :class:`atomx.models.PriceModel`
  * :class:`atomx.models.Timezone`
  * :class:`atomx.models.Zipcode`


1.5
---

- :meth:`atomx.Atomx.report` takes ``name`` parameter to name reports
- remove ``network`` scope in :meth:`atomx.Atomx.report` and add
  ``network_managed``, ``network_buy``, ``network_sell``
- Add new models:

  * :class:`atomx.models.AccountManager` (alias for `User`)
  * :class:`atomx.models.CampaignDebugReason`
  * :class:`atomx.models.CreativeAttribute`
  * :class:`atomx.models.PlacementType`
  * :class:`atomx.models.Visibility`

- Add :meth:`atomx.models.Report.save` to edit ``name`` and ``emails``
  of a :class:`atomx.models.Report`
- :meth:`atomx.Atomx.get` also accepts a model class or instance as resource argument.
  E.g.: ``atomx_api.get(atomx.models.Advertiser)`` or ``atomx_api.get(atomx.models.Advertiser(42))``
- Add :mod:`pickle` support for :mod:`atomx.models`.
- Save HTTP headers in :attr:`atomx.Atomx.last_response`.
- Add history support. :meth:`atomx.models.AtomxModel.history`.
- Use api version 3::

  * Fast reporting. No more polling if reports are ready.
  * Use authentication token in HTTP header instead of cookies.
  * Model and attribute changes


1.4
---

- Change default API version to `v2`
- If :meth:`atomx.Atomx.post` returns a list, auto-convert list of objects
  to a list of :mod:`atomx.models` models. (Useful for `POST` to the `/domains` endpoint)
- Add :class:`atomx.models.ScheduledReport`
- :meth:`atomx.Atomx.report` accepts ``when`` and ``interval`` to create a
  :class:`atomx.models.ScheduledReport`


1.3
---

- Add :meth:`atomx.Atomx.delete` to send a ``HTTP DELETE`` request to the api
- :meth:`atomx.Atomx.get` and :meth:`atomx.Atomx.delete` accept non-keyword arguments
  that are used to compute the final resource path
- Add `emails` parameter to :meth:`atomx.Atomx.report`
- Model attributes that are dates get automatically converted to a python :mod:`datetime`
- When saving a model, dates, sets and decimals get automatically converted
  to there json counterpart
- Add `save_response` parameter to :class:`atomx.Atomx` to save the response meta data
  of the last api call


1.2
---

- You can now remove model attributes with `del`
- Add :meth:`atomx.models.Report.csv` property that returns the report content as a list
- Save logged in user as `user` property to :class:`atomx.Atomx`
- Add network reports
- Try to determine report scope from user access rights if no scope was specified


1.1
---

- Fix: setup.py not working under some environments (`open` used wrong codec)
- Add SellerProfile model
- Add `offset` parameter to :meth:`atomx.models.Report.get`


1.0
---

- First release


tests.py
import pytest


@pytest.fixture(scope="session")
def atomx():
    from atomx import Atomx
    return Atomx('daniel@atomx.com', 'password', 'http://127.0.0.1:6543/v1/')


def test_limit(atomx):
    creatives = atomx.get('creatives', limit=5)
    assert len(creatives) == 5

def test_update(atomx):
    NEW_TITLE = 'test title'
    creative = atomx.get('creatives', id=5)
    creative.title = NEW_TITLE
    creative.update()
    creative = atomx.get('creatives', id=5)
    assert creative.title == NEW_TITLE


def test_save(atomx):
    from atomx.models import Profile
    profile = Profile(advertiser_id=23, name='test advertiser')
    profile.save()
    profile_new = atomx.get(Profile, id=profile.id)
    assert profile.name == profile_new.name


atomx/utils.py
import re
from atomx import models


def get_model_name(name):
    """Checks that :param:`name` is a valid model.
    Converts plural `name` to capitalized singular form,
    rstrips '_list' (e.g. `advertisers_list`)
    and '-' or '_' separation to camelCase.
    Returns the name of the model if found, False otherwise.
    E.g.

    >>> check_model_name('countries')
    "Country"
    >>> check_model_name('ADVERTISERS')
    "Advertiser"
    >>> check_model_name('conversion-pixels')
    "ConversionPixel"
    >>> check_model_name('operating_system')
    "OperatingSystem"
    >>> check_model_name('InvalidModel')
    False
    >>> check_model_name('advertisers_list')
    "Advertiser"

    :param str name: model name to convert
    :return: name of the model or False
    """
    if name.endswith('_list'):
        name = name[:-5]
    if '-' in name:
        model_name = ''.join(m.capitalize() for m in name.split('-'))
    elif '_' in name:
        model_name = ''.join(m.capitalize() for m in name.split('_'))
    else:
        model_name = name.capitalize()

    if model_name.endswith('ies'):  # change e.g. Countries to Countr*y*
        model_name = model_name[:-3]+'y'
    else:
        model_name = model_name.rstrip('s')
    if model_name in dir(models):
        return model_name
    return False


def get_attribute_model_name(attribute):
    """Checks if an attribute is a valid model like :func:`get_model_name`
    but also strips '_filter', '_include', '_exclude'

    :param str attribute: attribute name to convert
    :return: name of the model or False
    """
    return get_model_name(re.sub('(_filter|_include|_exclude)$', '', attribute))


def model_name_to_rest(name):
    """Gets a name of a :mod:`atomx.models` and transforms it in the
    resource name for the atomx api.
    E.g.::

        >>> assert model_name_to_rest('ConversionPixels') == 'conversion-pixels'
        >>> assert model_name_to_rest('OperatingSystem') == 'operating-system'
        >>> assert model_name_to_rest('Advertiser') == 'advertiser'

    :param str name: Name of the model to convert
    :return: :class:`str` with the transformed name.
    """
    r = name[0]
    for i in range(1, len(name)):
        if name[i].isupper():
            r += '-'
        r += name[i]
    return r.lower()


class _class_property(object):
    """Decorator to create @classmethod and @property"""
    def __init__(self, f):
        self.f = f

    def __get__(self, obj, owner):
        return self.f(owner)


atomx/exceptions.py
class NoSessionError(Exception):
    """Raised when a model from :mod:`.models` wants to ``create``/``update``
    but it doesn't have an :class:`atomx.Atomx` session yet.
    """
    pass

class InvalidCredentials(Exception):
    """Raised when trying to login with the wrong e-mail or password."""
    pass

class APIError(Exception):
    """Raised when the atomx api returns an error that is not caught otherwise."""
    pass

class ModelNotFoundError(Exception):
    """Raised when trying to (re-)load a model that is not in the api."""
    pass

class ReportNotReadyError(Exception):
    """Raised when requesting ``report.content`` but the report is not ready yet."""
    pass

class NoPandasInstalledError(Exception):
    """Raised when trying to access ``report.pandas`` without :mod:`pandas` installed."""
    pass


atomx/__init__.py
# -*- coding: utf-8 -*-

from datetime import (
    datetime,
    timedelta,
)
from inspect import isclass
import requests
from atomx.version import API_VERSION, VERSION
from atomx import models
from atomx.utils import (
    get_model_name,
    model_name_to_rest,
)
from atomx.exceptions import (
    APIError,
    ModelNotFoundError,
    InvalidCredentials,
    MissingArgumentError,
)


__title__ = 'atomx'
__version__ = VERSION
__author__ = 'Spot Media Solutions Sdn. Bhd.'
__copyright__ = 'Copyright 2015-2016 Spot Media Solutions Sdn. Bhd.'

API_ENDPOINT = 'https://api.atomx.com/{}'.format(API_VERSION)


class Atomx(object):
    """Interface for the api on api.atomx.com.

    To learn more about the api visit the
    `atomx wiki <https://wiki.atomx.com/api>`_

    :param str email: email address of your atomx user
    :param str password:  password of your atomx user
    :param str totp: 6 digit auth token if the account has 2-factor authentication enabled.
    :param str api_endpoint: url for connections to the api
        (defaults to `https://api.atomx.com/{API_VERSION}`)
    :param bool save_response: If `True` save the last api response meta info
        (without the resource payload) in :attr:`.Atomx.last_response`. (default: `True`)
    :return: :class:`.Atomx` session to interact with the api
    """
    def __init__(self, email, password, totp=None,
                 api_endpoint=API_ENDPOINT, save_response=True, expiration=None):
        self.auth_token = None
        self.user = None
        self.api_endpoint = api_endpoint.rstrip('/') + '/'
        self.save_response = save_response
        #: Contains the response of the last api call, if `save_response` was set `True`
        self.last_response = None
        self.login(email, password, totp, expiration)

    @property
    def _auth_header(self):
        if self.auth_token:
            return {'Authorization': 'Bearer ' + self.auth_token}

    def login(self, email, password, totp=None, expiration=None):
        """Gets new authentication token for user ``email``.

        This method is automatically called in :meth:`__init__` so
        you rarely have to call this method directly.

        :param str email: Email to use for login.
        :param str password: Password to use for login.
        :param str totp: 6 digit auth token if the account has 2-factor authentication enabled.
        :param int expiration: Number of seconds that the auth token should be valid. (optional)
        :return: None
        :raises: :class:`.exceptions.InvalidCredentials` if ``email``/``password`` is wrong
        """
        json = {'email': email, 'password': password}
        if totp:
            json['totp'] = str(totp)
        if expiration:
            json['expiration'] = expiration
        r = requests.post(self.api_endpoint + 'login', json=json)
        if not r.ok:
            if r.status_code == 401:
                raise InvalidCredentials
            raise APIError(r.json()['error'])
        self.auth_token = r.json()['auth_token']
        self.user = models.User(session=self, **r.json()['user'])

    def logout(self):
        """Removes authentication token from session."""
        self.auth_token = None
        self.user = None

    def search(self, query, index=None):
        """Search for ``query``.

        Returns a `dict` with all found results for:
        'Advertisers', 'Campaigns', 'Creatives', 'Placements', 'Publishers', 'Sites'.

        The resulting :mod:`.models` have only `id` and `name` loaded since that's
        what's returned from the api `/search` call, but attributes will be lazy loaded
        once you try to accessed them.
        Or you can just fetch everything with one api call with :meth:`.AtomxModel.reload`.

        Example::

            >>> atomx = Atomx('apiuser@example.com', 'password')
            >>> search_result = atomx.search('atomx')
            >>> assert 'campaigns' in search_result
            >>> campaign = search_result['campaigns'][0]
            >>> assert isinstance(campaign, models.Campaign)
            >>> # campaign has only `id` and `name` loaded but you
            >>> # can still access (lazy load) all attributes
            >>> assert isinstance(campaign.budget, float)
            >>> # or reload all attributes with one api call
            >>> campaign.reload()

        :param str query: keyword to search for.
        :param list index: :class:`str` or :class:`list` of the indexes you want to get returned.
            E.g. ``index=['campaigns', 'domains']``.
        :return: dict with list of :mod:`.models` as values
        """
        params = {'q': query}
        if index:
            if isinstance(index, list):
                index = ','.join(index)
            params['index'] = index
        r = requests.get(self.api_endpoint + 'search',
                         params=params,
                         headers=self._auth_header)
        r_json = r.json()
        if not r.ok:
            raise APIError(r_json['error'])
        search_result = r_json['search']

        if self.save_response:
            del r_json['search']
            self.last_response = r_json
            self.last_response['_headers'] = r.headers

        # convert publisher, creative dicts etc from search result to Atomx.model
        for m in search_result.keys():
            model_name = get_model_name(m)
            if model_name:
                search_result[m] = [getattr(models, model_name)(session=self, **v)
                                    for v in search_result[m]]
        return search_result

    def report(self, scope=None, groups=None, metrics=None, where=None,
               from_=None, to=None, daterange=None, timezone='UTC',
               emails=None, when=None, interval=None, name=None,
               sort=None, limit=None, offset=None, save=True, editable=False):
        """Create a report.

        See the `reporting atomx wiki <https://wiki.atomx.com/reporting>`_
        for details about parameters and available groups, metrics.

        :param str scope: Specifies the report type. Should be one of:
            'advertiser', 'publisher', 'inventory', 'dsp',
            'network_managed', 'network_buy', 'network_sell'.
            If undefined it tries to determine the `scope` automatically based
            on the access rights of the api user.
        :param list groups: columns to group by.
        :param list metrics: columns to sum on.
        :param list where: is a list of expression lists.
            An expression list is in the form of ``[column, op, value]``:

                - ``column`` can be any of the ``groups`` or ``metrics`` parameter columns.
                - ``op`` can be any of ``==``, ``!=``, ``<``, ``>``, ``in`` or ``not in`` as a string.
                - ``value`` is either a number or in case of ``in``
                  and ``not in`` a list of numbers.

        :param datetime.datetime from_: :class:`datetime.datetime` where the report
            should start (inclusive). (Defaults to last week)
        :param datetime.datetime to: :class:`datetime.datetime` where the report
            should end (exclusive). (Defaults to `datetime.now()` if undefined)
        :param str daterange: Use :param:`daterange` to automatically set the reports
        `from` and `to` parameters relativ to the current date.
        Both :param:`from_` and :param:`to` have to be ``None`` for it.
        Dateranges are: today, yesterday, last7days, last14days, last30days, monthtodate,
                        lastmonth, yeartodate, lifetime. (Defaults to ``None``)
        :param str timezone:  Timezone used for all times. (Defaults to `UTC`)
            For a supported list see https://wiki.atomx.com/timezones
        :param emails: One or multiple email addresses that should get
            notified once the report is finished and ready to download.
        :type emails: str or list
        :param str when: When should the scheduled report run. (daily, monthly, monday-sunday)
        :param str interval: Time period included in the scheduled report ('N days' or 'N month')
        :param str name: Optional name for the report.
        :param str or list sort: List of columns to sort by.
        :param int limit: Number of rows to return
        :param int offset: Number of rows to skip.
        :param bool save: Should the report appear in the users report history (defaults to `True`).
        :param bool editable: Should other users be able to change the date range of this report.
        :return: A :class:`atomx.models.Report` model
        """
        report_json = {'timezone': timezone, 'save': save, 'editable': editable}

        if name:
            report_json['name'] = name
        if groups:
            report_json['groups'] = groups
        if metrics:
            report_json['metrics'] = metrics
        elif not groups:
            raise MissingArgumentError('Either `groups` or `metrics` have to be set.')

        if scope is None:
            user = self.user
            if len(user.networks) > 0:
                pass  # user has network access so could be any report (leave scope as None)
            elif len(user.publishers) > 0 and len(user.advertisers) == 0:
                scope = 'publishers'
            elif len(user.advertisers) > 0 and len(user.publishers) == 0:
                scope = 'advertisers'

            if scope is None:
                raise MissingArgumentError('Unable to detect scope automatically. '
                                           'Please set `scope` parameter.')
        report_json['scope'] = scope

        if where:
            report_json['where'] = where

        if when and interval:  # scheduled report
            report_json['when'] = when
            report_json['interval'] = interval

        elif not from_ and not to and daterange:  # Rolling report
            report_json['daterange'] = daterange
        else:  # Normal report
            if from_ is None:
                from_ = datetime.now() - timedelta(days=7)
            if isinstance(from_, datetime):
                report_json['from'] = from_.strftime("%Y-%m-%d %H:00:00")
            else:
                report_json['from'] = from_

            if to is None:
                to = datetime.now()
            if isinstance(to, datetime):
                report_json['to'] = to.strftime("%Y-%m-%d %H:00:00")
            else:
                report_json['to'] = to

        if emails:
            if not isinstance(emails, list):
                emails = [emails]
            report_json['emails'] = emails

        params = {}
        if limit:
            params['limit'] = limit
        if offset:
            params['offset'] = offset
        if sort:
            if isinstance(sort, list):
                sort = ','.join(sort)
            params['sort'] = sort
        r = requests.post(self.api_endpoint + 'report',
                          params=params, json=report_json, headers=self._auth_header)
        r_json = r.json()
        if not r.ok:
            raise APIError(r_json['error'])
        report = r_json['report']

        if self.save_response:
            del r_json['report']
            self.last_response = r_json
            self.last_response['_headers'] = r.headers

        return models.Report(session=self, **report)

    def get(self, resource, *args, **kwargs):
        """Returns a list of models from :mod:`.models` if you query for
        multiple models or a single instance of a model from :mod:`.models`
        if you query for a specific `id`

        :param str resource: Specify the resource to get from the atomx api.

            Examples:

            Query all advertisers::

                >>> atomx = Atomx('apiuser@example.com', 'password')
                >>> advertisers = atomx.get('advertisers')
                >>> assert isinstance(advertisers, list)
                >>> assert isinstance(advertisers[0], atomx.models.Advertiser)

            Get publisher with id 23::

                >>> publisher = atomx.get('publisher/23')
                >>>> # or get the same publisher using the id as parameter
                >>> publisher = atomx.get('publisher', 23)
                >>>> # or use an atomx model
                >>> publisher = atomx.get(atomx.models.Publisher(23))
                >>> assert publisher.id == 23
                >>> assert isinstance(publisher, atomx.models.Publisher)

            Get all profiles for advertiser 42::

                >>> profiles = atomx.get('advertiser/42/profiles')
                >>> assert isinstance(profiles, list)
                >>> assert isinstance(profiles[0], atomx.models.Profile)
                >>> assert profiles[0].advertiser.id == 42

        :param args: All non-keyword arguments will get used to compute the ``resource``.
            This makes it easier if you want to work with a variable resource path.

            .. code-block:: python

                advertiser_id = 42
                attribute = 'profiles'
                profiles = atomx.get('advertiser', advertiser_id, attribute)
                # is equivalent to atomx.get('advertiser/42/profiles')

        :param kwargs: Any argument is passed as URL parameter to the respective api endpoint.
            See `API URL Parameters <https://wiki.atomx.com/api#url_parameters>`_
            in the wiki.

            Example:
            Get the first 20 domains that contain ``atom``::

                >>> atom_domains = atomx.get('domains', hostname='*atom*', limit=20)
                >>> assert len(atom_domains) == 20
                >>> assert 'atom' in atom_domains[1].hostname

        :return: a class from :mod:`.models` or a list of models depending on param `resource`
        """
        if isclass(resource) and issubclass(resource, models.AtomxModel):
            resource = resource._resource_name
        elif hasattr(resource, '_resource_name'):
            resource_path = resource._resource_name
            if hasattr(resource, 'id'):
                resource_path += '/' + str(resource.id)
            resource = resource_path
        else:
            resource = resource.strip('/')
        for a in args:
            resource += '/' + str(a)
        r = requests.get(self.api_endpoint + resource, params=kwargs, headers=self._auth_header)
        if not r.ok:
            raise APIError(r.json()['error'])

        r_json = r.json()
        model_name = r_json['resource']
        res = r_json[model_name]
        if self.save_response:
            del r_json[model_name]
            self.last_response = r_json
            self.last_response['_headers'] = r.headers
        model = get_model_name(model_name)
        if model and res:
            if isinstance(res, list):
                return [getattr(models, model)(session=self, **m) for m in res]
            return getattr(models, model)(session=self, **res)
        elif model_name == 'reporting':  # special case for `/reports` status
            return {
                'reports': [models.Report(session=self, **m) for m in res['reports']],
                'scheduled': [models.Report(session=self, **m) for m in res['scheduled']]
            }
        return res

    def post(self, resource, json, **kwargs):
        """Send HTTP POST to ``resource`` with ``json`` content.

        Used by :meth:`.models.AtomxModel.create`.

        :param resource: Name of the resource to `POST` to.
        :param json: Content of the `POST` request.
        :param kwargs: URL Parameters of the request.
        :return: :class:`dict` with the newly created resource.
        """
        r = requests.post(self.api_endpoint + resource.strip('/'),
                          json=json, params=kwargs, headers=self._auth_header)
        r_json = r.json()
        if not r.ok:
            raise APIError(r_json['error'])
        model_name = r_json['resource']
        res = r_json[model_name]
        if self.save_response:
            del r_json[model_name]
            self.last_response = r_json
            self.last_response['_headers'] = r.headers
        model = get_model_name(model_name)
        if model and isinstance(res, list):
            return [getattr(models, model)(session=self, **m) for m in res]
        return res

    def put(self, resource, id, json, **kwargs):
        """Send HTTP PUT to ``resource``/``id`` with ``json`` content.

        Used by :meth:`.models.AtomxModel.save`.

        :param resource: Name of the resource to `PUT` to.
        :param id: Id of the resource you want to modify
        :param json: Content of the `PUT` request.
        :param kwargs: URL Parameters of the request.
        :return: :class:`dict` with the modified resource.
        """
        r = requests.put(self.api_endpoint + resource.strip('/') + '/' + str(id),
                             json=json, params=kwargs, headers=self._auth_header)
        r_json = r.json()
        if not r.ok:
            raise APIError(r_json['error'])
        model_name = r_json['resource']
        res = r_json[model_name]
        if self.save_response:
            del r_json[model_name]
            self.last_response = r_json
            self.last_response['_headers'] = r.headers
        return res

    def delete(self, resource, *args, **kwargs):
        """Send HTTP DELETE to ``resource``.

        :param resource: Name of the resource to `DELETE`.
        :param args: All non-keyword arguments will be used to compute the final ``resource``.
        :param kwargs: Optional keyword arguments will be passed as query string to the
            delete request.
        :return: message or resource returned by the api.
        """
        resource = resource.strip('/')
        for a in args:
            resource += '/' + str(a)
        r = requests.delete(self.api_endpoint + resource, params=kwargs, headers=self._auth_header)
        r_json = r.json()
        if not r.ok:
            raise APIError(r_json['error'])
        model_name = r_json['resource']
        res = r_json[model_name]
        if self.save_response:
            del r_json[model_name]
            self.last_response = r_json
            self.last_response['_headers'] = r.headers
        return res

    def save(self, model):
        """Alias for :meth:`.models.AtomxModel.save` with `session` argument."""
        return model.save(self)

    def create(self, model):
        """Alias for :meth:`.models.AtomxModel.create` with `session` argument."""
        return model.create(self)

    def remove(self, model):
        """Alias for :meth:`.models.AtomxModel.delete` with `session` argument."""
        return model.delete(self)
